### PR TITLE
fix hostname resolution logic on ec2/ecs when the hostname_fqdn flag is enabled

### PR DESCRIPTION
**Pull request by arbll**
#### Fix an FQDN x EC2 bug in the hostname resolution

Some background first:

Currently, in the [hostname resolution logic](https://github.com/DataDog/datadog-agent/blob/fce7e5877371932df129ebf6d93cbb722b2f3ead/pkg/util/hostname.go#L104-L245) we have a variable called `hostName` that is used as a way to store the "best hostname we could find until now". This is the variable we return at the end of the logic.

The interface `Provider`, used to define hostname providers and implemented by the ec2 and gce hostname providers accepts a string as a parameter.
https://github.com/DataDog/datadog-agent/blob/fce7e5877371932df129ebf6d93cbb722b2f3ead/pkg/util/hostname/providers.go#L9
This string is used in and **only in** the ec2 provider logic. We use it to check if the given hostname matches the default ec2 hostname pattern. If it does we return the instance id and if it does not an error.
https://github.com/DataDog/datadog-agent/blob/fce7e5877371932df129ebf6d93cbb722b2f3ead/pkg/util/ec2/ec2.go#L85-L93

A second variable, called `name` is almost always used to store temporarily hostname values:
https://github.com/DataDog/datadog-agent/blob/595154c7e120cd8b3db072388adca7b3cb89741e/pkg/util/hostname.go#L166
https://github.com/DataDog/datadog-agent/blob/595154c7e120cd8b3db072388adca7b3cb89741e/pkg/util/hostname.go#L181

While it would probably have made sense to use the `hostName` variable to check that the current selected hostname is a default ec2 hostname, name was used instead:
https://github.com/DataDog/datadog-agent/blob/595154c7e120cd8b3db072388adca7b3cb89741e/pkg/util/hostname.go#L199

Unfortunately in https://github.com/DataDog/datadog-agent/pull/1822 I failed to identify this detail and used a variable called `fqdn` to temporarily store the FQDN:
https://github.com/DataDog/datadog-agent/blob/595154c7e120cd8b3db072388adca7b3cb89741e/pkg/util/hostname.go#L153

This led to a situation where on ec2 hosts with the `hostname_fqdn` flag enabled `hostName` was equal to the default ec2 FQDN (`ip-...`) and `name` to an empty string. Since empty string is not a default ec2 hostname, we were not replacing this default ec2 FQDN by the instance ID.

#### Modifications in this PR

- Remove the string parameter of the Provider interface (now `type Provider func() (string, error)`). It makes more sense to me to express the ec2 default hostname logic at the highest level. All other providers were just ignoring this param.
- Use a different variable to store each temporary hostname values to avoid side effects.
- Use `hostName` to check if the current selected hostname is a default ec2 hostname.

### Motivation

A few clients were hit by the issue.

### Additional Notes

- A potential edge case breaking change: `name` is always set by the providers, even if an error is returned. If for some reason the last provider that uses `name` before the ec2 logic returns both a valid ec2 hostname and an error the behavior will change after this PR.
- As written in the release note this will only affect ec2 agents with the `hostname_fqdn` option enabled.
- We lack testing on this logic

### Testing

[ ] ec2 x fqdn
[ ] ec2 x !fqdn
[ ] ecs x fqdn
[ ] ecs x !fqdn

### CODE REVIEW

**Review by olivielpeau — COMMENTED** [review]
Thanks for the detailed PR description!

Made an important comment on the logic, and a few nits on the release note.

File: pkg/util/hostname.go [hunk 1]
```diff
@@ -196,23 +197,27 @@ func GetHostname() (string, error) {
 	// and the hostname is one of the default ones
 	if getEC2Hostname, found := hostname.ProviderCatalog["ec2"]; found {
 		log.Debug("GetHostname trying EC2 metadata...")
-		instanceID, err := getEC2Hostname(name)
-		if err == nil {
-			err = ValidHostname(instanceID)
+		if ec2.IsDefaultHostname(hostName) {
+			instanceID, err := getEC2Hostname()
```
[the remark this answers is not in the document]
Comment (arbll) [reply pkg/util/hostname.go]: Good catch. It should be fixed now. I'll sanity check both ec2 and ecs once the pipeline builds the image.

**Review by olivielpeau — DISMISSED** [review]
Can we evaluate how many users/hosts would be impacted by this change? So that we can send a heads-up to these users. Let's discuss on slack.

Code changes LGTM, made a couple of nits.

**Comment by arbll** [comment]
Was previously approved, fixed issues with the build

**Pull request merged**